### ISSUE TITLE
[Question] Why doesn't stack use `extra-deps` from custom package's `packages`?

### ISSUE THREAD

**Issue by vlatkoB** [commercialhaskell/stack#486]
Hi,

I have one custom package that all my projects use, StdImports, and it is building fine with this .yaml:
```
flags: {}
packages:
- '.'
extra-deps:
- SafeSemaphore-0.10.1
- rdtsc-1.3.0.1
resolver: lts-2.16
```
Any project that is using the above StdImports must specify StdImport's `extra-deps` as well. 

```
flags: {}
packages:
- '.'
- ../StdImports2
extra-deps:
- SafeSemaphore-0.10.1
- rdtsc-1.3.0.1
resolver: lts-2.16
```
Without it, it wouldn't build. Shouldn't `stack` read .yaml of custom package, and use any info it finds there for building?

Or, is there another way to fake shared sandbox, and install StdImports there so other projects can use it without to much hassle?

I tried to specify both `packages` and `extra-deps` in global `.stack/stack.yaml`, but `stack` doesn't seem to use those info.

br,
vlatko

**Comment by snoyberg** [comment]
I was thinking about behavior like this just yesterday. I'd say the right
thing to do would be to have an import rule in the yaml file to include
values in other locations. This would keep the explicit nature of
stack.yaml but give the composability desired.

**Comment by vlatkoB** [comment]
This is what I'd expect without reading docs:
1. Read `global` .yaml
2. Read `all .yamls` in custom locations (i.e. packages and any other sections) 
    specified in current .yaml
3. Read `current` .yaml

Each next step overrides previous, if the same info exists.

This way we could have, for example, a global resolver if local is not specified, so changing LTS on a machine level would be easy. Also, patched package that all projects might use, etc.

No problem for values that are changed in current .yaml, but there should be a syntax to disable/ignore some inherited stuff. Something like, 
```
extra-deps:
- ignore-global: SafeSemaphore-0.10.1
```

Are your thoughts along these lines?

**Comment by snoyberg** [comment]
No, not quite. Please have a look at the stack.yaml Wiki page, which explains how the current inherit behavior works. It is specifically *not* at project-level settings such as resolver, as that would trivially lead to things breaking on one machine and succeeding on another. My thought was that we could have something like:

```yaml
extra-deps:
- ...
include:
- ../foo/stack.yaml
- ../bar/stack.yaml
```

or the equivalent.

**Comment by vlatkoB** [comment]
I think we'll be loosing a lot of flexibility and introducing more manual work, so that a project can be built on another machine, or at another time.
I agree both indeed are of outmost importance and understand that is the problem `stack` is trying to solve, but maybe that can be done in another way.

If local packages are specified, the project wouldn't build on another machine as well.
That might be solved by, for example, instead of simple copying the project's dir, `stack` can provide a command, such as `stack project freeze` or `stack project pack` or something, and prepare the project for moving/storing (i.e. making it really standalone) by including all local info necessary for build: global .yaml, local packages, resolved local symlinks, maybe even downloaded URL content (so we can really build it if some server goes down in the future) , etc.

Or, to be simple and please both views, provide an explicit project-level .yaml option `please-do-inherit-what-you-can`? ;-)

**Comment by snoyberg** [comment]
I'm really not getting it though. Why is:

```yaml
include:
- ../other-package
```

any worse than:

```yaml
extra-deps:
- other-package
```

?

**Comment by vlatkoB** [comment]
No, no problem there. That's a solution with clear meaning. It's a solution for the initial problem.
My last post was more about project-level .yaml isolation, and the lack of inheritance which I find very useful.

**Comment by snoyberg** [comment]
I'm afraid I don't understand your last post then, can you clarify? As far as automatically grabbing extra config values from random stack.yaml files: it can be very confusing from a principle-of-least-surprise.

**Comment by vlatkoB** [comment]
Maybe I expanded a bit from the original problem, but I was rather surprised that `stack` is envisioned with no inheritance at all, if I understood correctly. I explained what I expected/assumed in my [second post](https://github.com/commercialhaskell/stack/issues/486#issuecomment-118078198), and a partial solution I'd like to see in my [third post](https://github.com/commercialhaskell/stack/issues/486#issuecomment-118090659) (which maybe is not a bad idea at all for archiving purposes).
Currently I'm trying to convert several of my existing projects to `stack` and `stack init` always fails, so I have to copy `stack.yaml` or create it by hand and do the same manual editing in every project's .yaml, most of which could have been avoided if there is some sort of inheritance. Maybe `stack init` could have a global config with some preferred options and to advise it what to try if it can't succeed?
But, I'm looking at it from my perspective, which is probably more limited than what you have to envision for all possible uses. :-)

Let's get back to your proposal. 
```yaml
include:
- ../foo/stack.yaml
```
Did you mean that would include just the definitions from .yaml, not the package itself (i.e. package itself should be specified one more time in `packages` section)?
If so, than it would be possible to do something like:
```yaml
include:
- ~/.stack/global.yaml
```
That would solve the inheritance problem in an explicit way.

**Comment by snoyberg** [comment]
I do feel compelled to ask: did you read the stack.yaml wiki page that I referenced earlier? It is quite explicit about what is and is not inherited.

**Comment by vlatkoB** [comment]
Yes, read it few times, but noticed only info on separation of config options, to project and non-project specific, but nothing on inheriting/including project-external config options.
By [principle-of-least-surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), I got surprised by UI. ;-)

However, I fail to see what in my last question made you wonder about my reading stack.yaml wiki page. The question was about details on <i>explicit nature of including values in other locations</i>, as you described your idea. Have we misunderstood each other?

**Comment by snoyberg** [comment]
It seems like the second paragraph on the page is pretty clear about how config file parsing works:

> The stack.yaml configuration options break down into project specific and non-project specific options. The latter can be specified in your global config (/etc/stack/config) and user config (~/.stack/stack.yaml), as well as in the project config. The former must be in the project config.

We can discuss what makes the most sense, but I'm not sure why reading that would ever imply that a project-specific setting (like the list of packages) would ever be inherited.

**Comment by vlatkoB** [comment]
Well, it's definitely a misunderstanding. 
You sound like I was discussing what's written on the wiki page and what it does imply and does not. But I was just trying to discuss about the best way on how to incorporate some external options into the .yaml file. That's the initial problem this issue was opened for, you had some thoughts about it, and than discussion went in a strange direction. :-(

I have nothing to say on the matter of wiki, but would very much like to reach some constructive ideas on the best way to handle the problem, which might lead to implementation.

This thread has become rather long, hard to follow and not very constructive, so I might suggest to close this one and open another. What do you think?

**Comment by snoyberg** [comment]
That sounds like a good plan.
